### ISSUE TITLE
Creating processing pipelines (and other extensions to the EmitterMixin)

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-utils#8]
Moved from https://github.com/ckeditor/ckeditor5-core/issues/200

**Comment by scofalik** [comment]
I find it very weird, that providing low value for `priority` setting when defining callback actually resulting in this callback being fired sooner. So basically, low `priority` value = sooner execution. This is counter-intuitive and stands against of what priority means. High priority means that something should be executed sooner. It is also weird while documenting code. I wanted to write that we set high priority for the event but the value was `1`...

**Comment by Reinmar** [comment]
Yeah, documenting this was always tricky, because high priority means lower values. We could consider changing this, but:

1. I think that negative priorities would look weirder than now. Now sorting is done in priority order, so you start thinking about this value as about "order" (this is actually a better argument name), so negative ones are fine. But negative priorities will be weird. And if we change the order (and make `0` the default, as we discussed) you will be using them much more often than now.
2. EmitterMixin is the only API that CKE5 inherited after CKE4... so at the same time you would expect that it will work like CKE4's API and... it doesn't matter, cause the API is totally different anyway.

**Comment by scofalik** [comment]
I like your train of thoughts on point 2. :)

It's hard to do something with negative values, that's true. Maybe we could document them and make some special handling for negative values? But that would be really weird when one just look at the code and not read the docs.

**Comment by Reinmar** [comment]
>  Maybe we could document them and make some special handling for negative values?

I'm not sure what special handling you meant, but there should be no special treatment for negative values. They just need to behave like other priorities. Any special behaviour would be highly confusing.

**Comment by pjasiun** [comment]
For me it was WTF when I started working with the CKEditor. I think no user get it at the begging that the smaller priority value is a higher priority. On the other hand if we change it now it will be a WTF for that small group of users who learned, probably the hard way, that our priorities are reversed.

Seems are not able to write a sentence where we call it "priority" ("high priority means lower values", "thinking about this value as about "order""), we we have 2 options: renamed it or change it. If we want to keep the logic we should start calling it `priorityValue`, `priorityOrder` or `order`.

On the other hand changing the logic would make it simpler to describe once and for all.

And about the negative values: in CKEditor 4 we have many times problem what is the highest priority value, for the code which should be executed at the very end, when all other listeners were called. It was pretty common problem, and found in the code `999`, `1000`, `9999`, and maybe some other values. Now we could make it simple so define that the priority `0` is the lowest priority for the code which should be executed at the very end.

**Comment by pjasiun** [comment]
Of course, then the default priority should be somewhere around 100 or 1000.

**Comment by Reinmar** [comment]
Another improvement that we could make is removing the `context` param which is not needed any more (since we have arrow functions and bind()`).

**Comment by Reinmar** [comment]
Gosh... we're still thinking what to do with these priorities. None of the options look good – switching order, renaming to `order`, `index` or `importance`. We also need to define the highest value and the lowest value and the default one... Turns out that every option can be confusing.

Anyway, the latest idea is to use flags like `'LOWEST'`, `'LOW'`, `'NORMAL'`, `'HIGH'`, '`HIGHEST'`. That has some nice advantages:

1. It's super clear what those values mean.
2. We have the boundary values and the default clearly defined.
3. We limit the options, so we should not end up with all kind of priorities from -2 to 99999 like in CKEditor 4. If you need a priority between one of those two subsequent values you should clearly consider splitting the event into more meaningful parts.

WDYT? +1?

**Comment by scofalik** [comment]
I don't know if it will be good for converters. In converters we might need bigger granuality, because some converters will overwrite "default" versions of converters etc. etc. I hope we will not run into more problems because of flags.

Other than that, I'd propose 'VERY_LOW' and 'VERY_HIGH' just to have a few more options and then go for it.

**Comment by fredck** [comment]
> Anyway, the latest idea is to use flags like 'LOWEST', 'LOW', 'NORMAL', 'HIGH', 'HIGHEST'.

Is there anything good if we allow to combine them? e.g `LOW + LOW` is lower that `LOW` alone. I don't think that one would use anything other than the standard 5 options, but this may help on the edge cases.

**Comment by scofalik** [comment]
We considered using FLAG + offset but more as a joke :D

**Comment by fredck** [comment]
> Other solution is to have those as not flags but const values and then you can add offsets.

Exactly... we not even have to document that weird use of it, but it would be supported anyway out of the box then.

**Comment by scofalik** [comment]
What I am afraid is overuse of this and overlapping, let's say that we split those priorities by 100 and then someone does LOW - 100. And we won't have the "lowest boundary" priority. IDK. Using flags is very simple and straightforward.

**Comment by Reinmar** [comment]
I'd rather go with a flag and offset :D. Flags will be strings so `'LOW' + 'LOW'` will give `'LOWLOW'` :P.

**Comment by fredck** [comment]
Ofc, we were talking about having these as constants pointing to integers, like  `LOWEST=-1000, LOW=-100, NORMAL=0, HIGH=100, HIGHEST=1000`. 

Offset is the same problem than we have today, but if we go witn integers we would in any case support them... but if you want to force to be the lowest, for any reason, `LOWEST+LOWEST` would do the job :) Again, we're just talking about super edge cases

**Comment by pjasiun** [comment]
The problem is where this integers should be put? What you should import to use it? `emittermixin.js`? Add them as static propertied of a class you mixin? Special `prorieties.js`? Stupid problem, but still a problem.

**Comment by scofalik** [comment]
That's why FLAG + offset seems reasonable, but I still think that we should try to be okay with like 7 flags.

**Comment by Reinmar** [comment]
Argh... there's one super irritating thing currently – no easy way to off an listener on destroy.

I have a `Buffer` class which has a method `_onChange`. This class listens to `Document#change` and calls that `_onChange` method. Now, how do I detach that class from the document? That obviously won't work:

```js
destroy() {
    this.document.off( 'change', this._onChange );
}
```

Because I added the listener using an arrow function (to preserve ctx and pass the right args).

The only option now is to either store that arrow function in a property, so we can later use it in `destroy()` or create an instance of `Emitter` like this:

```js

this._observer = Object.create( EmitterMixin );
this._observer.listenTo( document, 'change', () => this._onChange() );
```

So we can later do:

```js
this._observer.stopListening();
```

And it's more or less fine... but if `on()` was returning an object with an `off()` method (like in CKEditor 4) there would be no issue at all. So that's another thing on the list of possible improvements.

**Comment by Reinmar** [comment]
There's even more... https://github.com/ckeditor/ckeditor5/pull/183#discussion_r64540663

The tl;dr of that thread is that I'd rename the method to `observer()` so we can do `observer.observe()` which is what was for me intuitive.

I think that the amount of aspects mentioned in this ticket raises it priority quite high. We'll need to address all this soon.

**Comment by Reinmar** [comment]
We've got to restore this topic finally.

We had bazillion ideas what could be improved in the emitter mixin, but we weren't able to clarify the direction that we want to take with this feature.

The following aspects were mentioned:

1. Priorities.
2. `on()` vs `listenTo()` – that the latter should be used in most cases, so perhaps we can somehow promote it through a different name.
3. `Listener` vs `observer` ([see](https://github.com/ckeditor/ckeditor5-utils/issues/8#issuecomment-221518948)).
4. Returning values (and data pipelines).
5. Performance.

## Priorities

Fortunately, this one is most clear (at least to me). I propose to go with string values as proposed in https://github.com/ckeditor/ckeditor5-utils/issues/8#issuecomment-210017305 (with the difference that we use lower case now). The param name would be still `priority`.

We were discussing ability to specify `'lowest' - 1` as a priority for debugging purposes (when you need to set the priority precisely). That's not going to work ofc (you'll get `NaN`), but we can do it differently. The string values should be converted to numbers inside `on()`, but we can also accept numbers. `EmitterMixin` can expose numeric values of priorities through:

```js
import { listenerPriorities } from './emittermixin.js';

listenerPriorities.low; // e.g. -100

// To precisely set the listener's priority (for debugging or tests purposes only!):
foo.on( 'x', fn, listenerPriorities.low - 1 ); 

// How you'd normally use them:
foo.on( 'x', 'fn', 'low' );
```

## `on()` vs `listenTo()` and `listener` vs `observer`

I wouldn't touch this topic now. We'll be reviewing the usage of listeners when working on destroying the editor, so there's going to be some time for such considerations and we'll also know more then on what is a real problem.

## Returning values

We've handled this topic by passing special container objects to listeners. For now it's enough. Let's see later on once we'll have pipelines like the paste data processing. There's a chance that we won't need more than what we have now.

# Plan of action

1. Work on the priorities (in a new ticket).
2. Remove a milestone from this general issue and keep it open for future discussions.

**Comment by pjasiun** [comment]
> // To precisely set the listener's priority (for debugging or tests purposes only!):
> foo.on( 'x', fn, listenerPriorities.low - 1 );

I do not see it useful. For debugging purpose, you may know that 'low' is `10` and use 9. For other purposes you should not use `listenerPriorities.low - 1`, at least we should try to avoid it, at least for now. Also, it is still unclear for me: do `low - 1` is before or after `low`?

**Comment by Reinmar** [comment]
For debugging you can presume that low is 10. But for testing you can't. Tests should work in the future if those numbers change for any reason.

**Comment by pjasiun** [comment]
I do not think they will change. And still think we should not over-engineer it and do it the way others do it: https://developer.android.com/reference/android/app/Notification.html#PRIORITY_HIGH

**Comment by Reinmar** [comment]
> Also, it is still unclear for me: do low - 1 is before or after low?

As the implementer wants. It doesn't matter to me. We'll have to care very rarely in a very few tests.

**Comment by Reinmar** [comment]
> I do not think they will change. And still think we should not over-engineer it and do it the way others do it: https://developer.android.com/reference/android/app/Notification.html#PRIORITY_HIGH

Whatever, really. It's a matter of whether we'll export that object with priority values or no. I'm fine with both.
